File: app/src/main/java/com/ratajczykdev/inventoryapp/database/ProductListRecyclerAdapter.java
package com.ratajczykdev.inventoryapp.database;
//  TODO: move outside package

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.ratajczykdev.inventoryapp.R;

import java.util.List;
import java.util.Locale;

public class ProductListRecyclerAdapter extends RecyclerView.Adapter<ProductListRecyclerAdapter.ProductViewHolder> {

    private final LayoutInflater layoutInflater;
    private List<Product> productList;

    /**
     * ViewHolder class
     */
    class ProductViewHolder extends RecyclerView.ViewHolder {
        private final TextView textName;
        private final ImageView imagePhoto;
        private final TextView textPrice;
        private final TextView textQuantity;

        public ProductViewHolder(View itemView) {
            super(itemView);
            textName = itemView.findViewById(R.id.text_name);
            imagePhoto = itemView.findViewById(R.id.image_photo);
            textPrice = itemView.findViewById(R.id.text_price);
            textQuantity = itemView.findViewById(R.id.text_quantity);
        }
    }

    ProductListRecyclerAdapter(Context context) {
        layoutInflater = LayoutInflater.from(context);
    }

    @NonNull
    @Override
    public ProductViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View itemView = layoutInflater.inflate(R.layout.product_list_item, parent, false);
        return new ProductViewHolder(itemView);
    }

    @Override
    public void onBindViewHolder(@NonNull ProductViewHolder holder, int position) {
        if (productList != null) {
            Product currentProduct = productList.get(position);
            holder.textName.setText(currentProduct.getName());
            //  TODO: make photo set method
            holder.textPrice.setText(String.format(Locale.US, "%.2f", currentProduct.getPrice()));
            holder.textQuantity.setText(String.valueOf(currentProduct.getQuantity()));
        } else {
            //  TODO: add empty list view, delete this
            holder.textName.setText("No product available");
        }
    }

    @Override
    public int getItemCount() {
        if (productList != null) {
            return productList.size();
        } else {
            return 0;
        }
    }

    /**
     * Use of notifyDataSetChanged()
     */
    void setProducts(List<Product> productList) {
        this.productList = productList;
        notifyDataSetChanged();
    }

}

File: app/src/main/java/dev/ratajczyk/inventoryapp/catalog/ProductListRecyclerAdapter.kt
package dev.ratajczyk.inventoryapp.catalog

import android.content.Context
import android.content.Intent
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.squareup.picasso.Picasso
import dev.ratajczyk.inventoryapp.R
import dev.ratajczyk.inventoryapp.database.Product
import dev.ratajczyk.inventoryapp.detailandedit.ProductDetailActivity
import java.util.*

/**
 * Provides data from list of [Product] to [RecyclerView]
 *
 * @author Mikolaj Ratajczyk <mikolaj.ratajczyk(AT)gmail.com>
 */
class ProductListRecyclerAdapter(val context: Context) : RecyclerView.Adapter<ProductListRecyclerAdapter.ProductViewHolder>() {

    companion object {
        const val DATA_SELECTED_PRODUCT_ID = "DATA_SELECTED_PRODUCT_ID"
    }

    private val layoutInflater = LayoutInflater.from(context)
    var productList: List<Product> = emptyList()
        set(value) {
            field = value
            notifyDataSetChanged()
        }


    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ProductViewHolder {
        val itemView = layoutInflater.inflate(R.layout.product_list_item, parent, false)
        return ProductViewHolder(itemView)
    }

    override fun onBindViewHolder(holder: ProductViewHolder, position: Int) {
        val currentProduct = productList[position]
        holder.textName.text = currentProduct.name
        holder.textPrice.text = String.format(Locale.US, "%.2f", currentProduct.price)
        holder.textQuantity.text = currentProduct.quantity.toString()
        loadProductPhotoFromUri(holder, currentProduct)

        holder.listItemRootView.setOnClickListener {
            val intent = Intent(context, ProductDetailActivity::class.java)
            intent.putExtra(DATA_SELECTED_PRODUCT_ID, currentProduct.id.toString())
            context.startActivity(intent)
        }
    }

    /**
     * Helper method, it loads product photo to ImageView
     */
    private fun loadProductPhotoFromUri(holder: ProductViewHolder, currentProduct: Product) {
        Picasso.get()
                .load(currentProduct.photoUri)
                .placeholder(R.drawable.product_list_item_placeholder)
                .error(R.drawable.ic_error)
                .fit()
                .into(holder.imagePhoto)
    }

    override fun getItemCount(): Int {
        return productList.size
    }

    class ProductViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        internal val textName: TextView = itemView.findViewById(R.id.text_name)
        internal val imagePhoto: ImageView = itemView.findViewById(R.id.image_photo)
        internal val textPrice: TextView = itemView.findViewById(R.id.text_price)
        internal val textQuantity: TextView = itemView.findViewById(R.id.text_quantity)
        internal val listItemRootView: View = itemView.findViewById(R.id.product_list_item_root_linearlayout)
    }
}